File: Strohm/ios/Strohm/Sources/Strohm/ViewModels/SimpleViewModel.swift
import Foundation

open class SimpleViewModel<EntryType: ConstructableFromDictionary>: ViewModelBase<EntryType>, PropsHandler {

    public var data: EntryType {
        willSet {
            // https://stackoverflow.com/questions/57615920/published-property-wrapper-not-working-on-subclass-of-observableobject
            self.objectWillChange.send()
        }
    }

    public init(initialData: EntryType, propName: PropName, propPath: PropPath) {
        self.data = initialData
        super.init(propName: propName, propPath: propPath)
    }

    override func propsToData(props: Props) -> EntryType? {
        guard let rawData = props[self.propName] as? [String:Any],
           let data = EntryType.init(from: rawData) else {
            return nil
        }

        print("Received entry: ", data)
        return data
    }

    override func store(data: EntryType) {
        self.data = data
    }
}

File: Strohm/android/strohm/src/main/java/dev/strohmnative/strohm/viewmodels/SimpleViewModel.kt
package dev.strohmnative.strohm.viewmodels

import android.util.Log
import dev.strohmnative.strohm.*

open class SimpleViewModel<EntryType>(
    initialData: EntryType,
    propName: PropName,
    propPath: PropPath,
    private val instanceFactory: ConstructableFromDictionary<EntryType>
): ViewModelBase<EntryType>(initialData, propName, propPath) {
    override fun propsToData(props: Props): EntryType? {
        val m = props[this.propName] as? Map<*, *> ?: return null
        val rawData = m.asMapOfType<PropName, Props>() ?: return null
        var data: EntryType? = instanceFactory.createFromDict(rawData) ?: return null

        Log.d("dev.strohmnative.strohm", "Received entry: $data")
        return data
    }
}